Improve  Unix compliance `gh repo set-default`

thecaffeinedev opened the pull request:
Fixes #9398 

I have manually tested the command in various scenarios to ensure correct behavior.

### Changes:
1. Error messages are now printed to stderr instead of stdout when no default repository is set.
2. The command  `gh repo set-default --view` now returns a non-zero exit code (SilentError) in view mode when no default repository is set.
3. Updated and added tests to cover the new error handling behavior

I am open to any feedback or additional changes that may be needed.

thecaffeinedev committed:
Improve  Unix compliance

Signed-off-by: Prabhat <iprabhatdev@gmail.com>

jtmcg commented on pkg/cmd/repo/setdefault/setdefault.go:
```diff
@@ -121,12 +121,13 @@ func setDefaultRun(opts *SetDefaultOptions) error {
+			return nil
 		} else if opts.IO.IsStdoutTTY() {
-			fmt.Fprintln(opts.IO.Out, "no default repository has been set; use `gh repo set-default` to select one")
+			fmt.Fprintln(opts.IO.ErrOut, "no default repository has been set; use `gh repo set-default` to select one")
+			return cmdutil.SilentError
```
I think we'd prefer to just return an error here
```suggestion
			return fmt.Errorf("no default repository has been set; use `gh repo set-default` to select one")
```

jtmcg commented on pkg/cmd/repo/setdefault/setdefault.go:
```diff
@@ -121,12 +121,13 @@ func setDefaultRun(opts *SetDefaultOptions) error {
+			return nil
 		} else if opts.IO.IsStdoutTTY() {
-			fmt.Fprintln(opts.IO.Out, "no default repository has been set; use `gh repo set-default` to select one")
+			fmt.Fprintln(opts.IO.ErrOut, "no default repository has been set; use `gh repo set-default` to select one")
+			return cmdutil.SilentError
 		}
-		return nil
+		return cmdutil.SilentError
```
I think we want to continue to return `nil`, here. In fact, I think we can remove the `else` entirely:

The command isn't failing, so we aren't returning an error. That would introduce a breaking change. Instead, we are routing the messaging to `stderr` instead of `stdout` so that we log to the user _why_ this isn't returning anything without breaking any piping or automation experiences.

andyfeller replied:
> _The command isn't failing, so we aren't returning an error._

To dig into this deeper, `stderr` isn't solely used for telling users about errors; CLIs trying to avoid breaking Unix command output / input piping use it for a whole variety of reasons.

Additionally, `gh` has a predictable approach to determining the base repo in various situations while giving users the ability to override this logic using `gh repo set-default`.  It isn't an error if the user has not explicitly overrode the base repo, so this should continue returning `0`.

jtmcg commented:
Thanks for the submission, @thecaffeinedev! It occurs to me that the ask in the issue may not have been super clear 😅 That's on me! I'll update #9398 with a comment to clearly outline what the ask is (it'll look quite similar to my comment on the PR).

thecaffeinedev commented:
> Thanks for the submission, @thecaffeinedev! It occurs to me that the ask in the issue may not have been super clear 😅 That's on me! I'll update #9398 with a comment to clearly outline what the ask is (it'll look quite similar to my comment on the PR).

Thank you for the feedback and clarification, @jtmcg. I appreciate you taking the time to review the PR. 

Do you want me to update my PR to implement [this specific change?](https://github.com/cli/cli/issues/9398#issuecomment-2274009879)

jtmcg commented:
> > Thanks for the submission, @thecaffeinedev! It occurs to me that the ask in the issue may not have been super clear 😅 That's on me! I'll update #9398 with a comment to clearly outline what the ask is (it'll look quite similar to my comment on the PR).
> 
> Thank you for the feedback and clarification, @jtmcg. I appreciate you taking the time to review the PR.
> 
> Do you want me to update my PR to implement [this specific change?](https://github.com/cli/cli/issues/9398#issuecomment-2274009879)

That would be great! Thanks again for the contribution 🙌

thecaffeinedev committed:
fix

thecaffeinedev commented:
> > > Thanks for the submission, @thecaffeinedev! It occurs to me that the ask in the issue may not have been super clear 😅 That's on me! I'll update #9398 with a comment to clearly outline what the ask is (it'll look quite similar to my comment on the PR).
> > 
> > 
> > Thank you for the feedback and clarification, @jtmcg. I appreciate you taking the time to review the PR.
> > Do you want me to update my PR to implement [this specific change?](https://github.com/cli/cli/issues/9398#issuecomment-2274009879)
> 
> That would be great! Thanks again for the contribution 🙌

I have updated the code. Also, I initially pushed a commit with a message starting with "fix". After realizing that this might not be ideal, I attempted to change the commit message using a force push. However, while signing the commit, it changed my message, and I was unable to modify the existing commit successfully.

Hope that's fine.

jtmcg commented on pkg/cmd/repo/setdefault/setdefault_test.go:
```diff
@@ -498,7 +500,7 @@ func TestDefaultRun(t *testing.T) {
 			}
 			assert.NoError(t, err)
-			assert.Equal(t, tt.wantStdout, stdout.String())
+			assert.Equal(t, tt.wantStdout, stdout.String()+stderr.String())
```
Changing the assert this way doesn't actually cover the change we're making. It will pass or fail regardless of whether or not we are printing to `stdout` or `stderr` on line 125 of `setdefault.go`. 

That's because, for this change, `stdout` is `""` while `stderr` is `"no default repository has been set; use `gh repo set-default` to select one"`. The opposite is true before we make this change. Either way, `stdout.String()+stderr.String()` will return `"no default repository has been set; use `gh repo set-default` to select one"`, so we aren't actually know if the message came from `stdout` or `stderr`.

I'd recommend extracting the logic to check `stdout` & `stderr` to their own `asserts`

jtmcg commented on pkg/cmd/repo/setdefault/setdefault_test.go:
```diff
@@ -179,13 +179,15 @@ func TestDefaultRun(t *testing.T) {
 					Repo:   repo1,
 				},
 			},
+			wantStdout: "no default repository has been set; use `gh repo set-default` to select one\n",
```
Since we are explicitly moving this from `stdout` to `stderr`, I think I'd prefer to see this as `wantStderr` rather than `wantStdout`

jtmcg reviewed (changes requested):
Right on, thanks for updating this! The logic is correct, but the tests need some updating. I've dropped a couple of comments on the test file to explain my thinking. Let me know if you have any questions.

jtmcg commented:
> > > > Thanks for the submission, @thecaffeinedev! It occurs to me that the ask in the issue may not have been super clear 😅 That's on me! I'll update #9398 with a comment to clearly outline what the ask is (it'll look quite similar to my comment on the PR).
> > > 
> > > 
> > > Thank you for the feedback and clarification, @jtmcg. I appreciate you taking the time to review the PR.
> > > Do you want me to update my PR to implement [this specific change?](https://github.com/cli/cli/issues/9398#issuecomment-2274009879)
> > 
> > 
> > That would be great! Thanks again for the contribution 🙌
> 
> I have updated the code. Also, I initially pushed a commit with a message starting with "fix". After realizing that this might not be ideal, I attempted to change the commit message using a force push. However, while signing the commit, it changed my message, and I was unable to modify the existing commit successfully.
> 
> Hope that's fine.

I can help clean up the git history before we merge. I'll try to capture my steps for you so you can see how I did it and try it on your own, too. 

Thanks for calling it out 👍

thecaffeinedev committed:
Fix: Print no default repository message to stderr

thecaffeinedev commented:
> Right on, thanks for updating this! The logic is correct, but the tests need some updating. I've dropped a couple of comments on the test file to explain my thinking. Let me know if you have any questions.

Thank you for your feedback. I have made the following changes to address your comments.

- I modified the `setdefault_test.go` file to check stdout and stderr output separately.
- Updated the relevant test cases to expect the message on stderr, using a new `wantStderr` field in the test struct.

Please let me know if you would like me to make any further adjustments.

jtmcg commented on pkg/cmd/repo/setdefault/setdefault_test.go:
```diff
@@ -499,6 +504,7 @@ func TestDefaultRun(t *testing.T) {
 			}
 			assert.NoError(t, err)
 			assert.Equal(t, tt.wantStdout, stdout.String())
+			assert.Equal(t, tt.wantStderr, stderr.String())
```
Last thing here: I think it'd be helpful to separate these with a conditional:

```suggestion
			if tt.wantStdout != "" {
			    assert.Equal(t, tt.wantStdout, stdout.String())
			} else {
			    assert.Equal(t, tt.wantStderr, stderr.String())
			}
```

This allows us to get only one string matching assert error per test case, meaning it's easier to understand what in the test is failing (`Stderr` or `Stdout`). Right now, we're getting two error outputs that are mirrors of one another, making understanding which part of the code is failing a bit more difficult.

jtmcg commented on pkg/cmd/repo/setdefault/setdefault_test.go:
```diff
@@ -175,17 +176,21 @@ func TestDefaultRun(t *testing.T) {
 				},
 			},
-			wantStdout: "no default repository has been set; use `gh repo set-default` to select one\n",
+			wantStdout: "",
```
We can remove this empty string

jtmcg commented on pkg/cmd/repo/setdefault/setdefault_test.go:
```diff
@@ -175,17 +176,21 @@ func TestDefaultRun(t *testing.T) {
 					Repo:   repo1,
 				},
 			},
+			wantStdout: "",
```
We can remove this empty string

thecaffeinedev replied:
Thank you for the additional feedback. I have updated the code based on your suggestions.

jtmcg reviewed (changes requested):
Super close. When I ran this the error outputs were a bit confusing so I've commented accordingly. Thanks for keeping up on it and responding so quickly!

thecaffeinedev committed:
Merge branch 'cli:trunk' into improve-set-default-unix-compliance

thecaffeinedev committed:
Refine setdefault tests for clearer stdout/stderr assertions

jtmcg reviewed (approved):
Looks good! Thanks again for the contribution 🍻
